GovernedLotteryHook.sol

```diff
@@ -0,0 +1,188 @@
+                proposalId: proposals.length,
+                description: description,
+                newSwapFeePercentage: newSwapFeePercentage,
+                newLuckyNumber: newLuckyNumber,
+                votesFor: 0,
+                votesAgainst: 0,
+                votingDeadline: block.timestamp + 7 days
+            })
+        );
+
+        emit ProposalCreated(proposals.length - 1, description);
+    }
+
+    // Vote on an active proposal
+    function voteOnProposal(uint256 proposalId, bool support) external {
+        require(block.timestamp <= proposals[proposalId].votingDeadline, "Voting period is over");
+        require(!hasVoted[proposalId][msg.sender], "You have already voted");
+
+        if (support) {
+            proposals[proposalId].votesFor += 1;
+        } else {
+            proposals[proposalId].votesAgainst += 1;
+        }
+
+        hasVoted[proposalId][msg.sender] = true;
+        emit VoteCast(proposalId, msg.sender, support);
+    }
+
+    // Implement the proposal if it has more votes for than against
+    function implementProposal(uint256 proposalId) external onlyOwner {
+        Proposal storage proposal = proposals[proposalId];
+        require(block.timestamp > proposal.votingDeadline, "Voting period not ended");
+
+        if (proposal.votesFor > proposal.votesAgainst) {
+            hookSwapFeePercentage = proposal.newSwapFeePercentage;
+            LUCKY_NUMBER = proposal.newLuckyNumber;
+
+            emit ProposalImplemented(proposalId, proposal.newSwapFeePercentage, proposal.newLuckyNumber);
+        }
+    }
+
+    // Lottery logic (onAfterSwap remains unchanged for the most part)
+    function onAfterSwap(
+        AfterSwapParams calldata params
+    ) public override onlyVault returns (bool success, uint256 hookAdjustedAmountCalculatedRaw) {
+        uint8 drawnNumber;
+        if (params.router == _trustedRouter) {
+            drawnNumber = _getRandomNumber();
+        }
+
+        _counter++;
+
+        hookAdjustedAmountCalculatedRaw = params.amountCalculatedRaw;
+        if (hookSwapFeePercentage > 0) {
+            uint256 hookFee = params.amountCalculatedRaw.mulDown(hookSwapFeePercentage);
+            if (params.kind == SwapKind.EXACT_IN) {
+                uint256 feeToPay = _chargeFeeOrPayWinner(params.router, drawnNumber, params.tokenOut, hookFee);
+                if (feeToPay > 0) {
+                    hookAdjustedAmountCalculatedRaw -= feeToPay;
+                }
+            } else {
+                uint256 feeToPay = _chargeFeeOrPayWinner(params.router, drawnNumber, params.tokenIn, hookFee);
+                if (feeToPay > 0) {
+                    hookAdjustedAmountCalculatedRaw += feeToPay;
+                }
+            }
+        }
+        return (true, hookAdjustedAmountCalculatedRaw);
+    }
+
+    // Function to set swap fee (can also be changed by governance)
+    function setHookSwapFeePercentage(uint64 swapFeePercentage) external onlyOwner {
```

Comment:
As written, only the owner can set the fee (which would need to be validated in production). And if they can just set the fee directly, the voting doesn't really mean anything. (Though it already doesn't, if only the owner can propose any changes.)

If you want governance to be able to set it, you would have to implement that (e.g., derive from `SingletonAuthentication` and use the authenticate modifier). Then it could be set by Balancer governance *or* through voting. (Balancer governance would be unlikely to get involved with a third party hook like this, though.)